src/framework/components/element/markup.js

```diff
@@ -0,0 +1,417 @@
+            }
+            return (this._mode === "text") ? this._text() : this._tag();
+        },
+
+        // read text block until eof or start of tag
+        _text: function () {
+            while (true) {
+                switch (this._cur) {
+                    case null:
+                        // reached end of input
+                        return (this._buf.length > 0) ? this.TEXT_TOKEN : this.EOF_TOKEN;
+                    case "[":
+                        // start of tag mode
+                        this._mode = "tag";
+                        return (this._buf.length > 0) ? this.TEXT_TOKEN : this._tag();
+                    case "\\":
+                        // handle escape sequence
+                        this._next();           // skip \
+                        switch (this._cur) {
+                            case "[":
+                                this._store();
+                                break;
+                            default:
+                                // if we don't recognize the escape sequence, output
+                                // the slash without interpretation and continue
+                                this._output("\\");
+                                break;
+                        }
+                        break;
+                    default:
+                        this._store();
+                        break;
+                }
+            }
+        },
+
+        // read tag block
+        _tag: function () {
+            while (true) {
+                switch (this._cur) {
+                    case null:
+                        this._error = "unexpected end of input reading tag";
+                        return this.ERROR_TOKEN;
+                    case "[":
+                        this._store();
+                        return this.OPEN_BRACKET_TOKEN;
+                    case "]":
+                        this._store();
+                        this._mode = "text";
+                        return this.CLOSE_BRACKET_TOKEN;
+                    case "=":
+                        this._store();
+                        return this.EQUALS_TOKEN;
+                    case " ":
+                    case "\t":
+                    case "\n":
+                    case "\r":
+                    case "\v":
+                    case "\f":
+                        return this._whitespace();
+                    case "\"":
+                        return this._string();
+                    default:
+                        if (!this._isIdentifierSymbol(this._cur)) {
+                            this._error = "unrecognized character";
+                            return this.ERROR_TOKEN;
+                        }
+                        return this._identifier();
+                }
+            }
+        },
+
+        _whitespace: function () {
+            this._store();
+            while (" \t\n\r\v\f".indexOf(this._cur) !== -1) {
```

Comment:
Maybe we should make this string a constant to avoid creating it on each `_whitespace` call?

Reply:
Do you perhaps have a link to documentation on the performance characteristics of a case like this? I would have expected constants embedded in code like this to be rather fast (but of course I'm used to c++).

Reply:
I don't have a link but I *think* that this string will be recreated everytime you call the _whitespace function and also possibly on every iteration of the while loop. I don't think it's slow but as a principle it'd be good to avoid allocating anything at runtime to avoid GC spikes.